File: programmers/new/level2/문자열_압축.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = Integer.MAX_VALUE;
        for(int i = 1; i <= s.length(); ++i) {
        	int shortenS = shorten(s, i);
        	if(shortenS < answer) {
        		answer = shortenS;
        	}
        }
        return answer;
    }
    int shorten(String s, int size) {
    	int currentCount = 1;
    	StringBuilder sb = new  StringBuilder();
    	int ni = 0;
    	for(int i = 0; i < s.length()-size; i+=size) {
    		ni = i+size;
    		boolean equal = true;
    		for(int j = 0; j < size; ++j) {
    			if(ni+j == s.length()) {
    				equal = false;
    				break;
    			}
    			if(s.charAt(i+j) != s.charAt(ni+j)) {
    				equal = false;
    				break;
    			}
    		}
    		if(equal) {
    			currentCount++;
    		}else {
    			if(currentCount>1) {
    				sb.append(currentCount);
    			}
    			for(int j = 0; j < size; ++j) {
    				sb.append(s.charAt(i+j));
    			}
    			currentCount = 1;
    		}
    	}
    	if(currentCount>1) {
			sb.append(currentCount);
		}
		for(; ni < s.length(); ++ni) {
			sb.append(s.charAt(ni));
		}
		return sb.length();
    }
}
public class 문자열_압축 {
	public static void main(String[] args) throws Exception{
		String s = "x";
		Solution sol = new Solution();
		System.out.println(sol.solution(s));
	}
}

File: programmers/js/new/level2/문자열_압축.js
function solution(s) {
    var answer = 1000;
    for (let i = 1; i <= s.length; ++i) {
        const compress = shorten(s, i);
        if (compress.length < answer) {
            answer = compress.length;
        }
    }
    return answer;
}

function shorten(s, n) {
    let result = "";
    let current = "";
    let count = 1;
    for (let i = 0; i < n; ++i) {
        current += s.charAt(i);
    }
    for (let i = n; i < s.length; i += n) {
        let next = "";
        for (let j = i; j < i + n; ++j) {
            next += s.charAt(j);
        }
        if (current === next) {
            count++;
        } else {
            result += current + (count > 1 ? count : "");
            current = next;
            count = 1;
        }
    }
    result += current + (count > 1 ? count : "");
    return result;
}
